File: fuzz/signal_edge.c
/*
 * fuzz/signal_edge.c — Edge-based signal tracking + corpus minimization.
 *
 * Signal = hash(PC ^ prev_PC) for each consecutive PC pair in KCOV trace.
 * Corpus keeps only programs that contribute at least one unique edge.
 * Minimization removes programs whose edges are all covered by others.
 */
#include "signal_edge.h"

#include <stdlib.h>
#include <string.h>
#include <stdio.h>

/* ─── Hash function (fast, good distribution) ─────────────────────────────── */

static inline uint32_t edge_hash(unsigned long pc, unsigned long prev_pc)
{
	uint64_t h = (uint64_t)(pc ^ prev_pc);
	h ^= h >> 33;
	h *= 0xff51afd7ed558ccdULL;
	h ^= h >> 33;
	h *= 0xc4ceb9fe1a85ec53ULL;
	h ^= h >> 33;
	return (uint32_t)(h & (SIGNAL_MAP_SIZE - 1));
}

/* ─── Edge signal from KCOV PCs ───────────────────────────────────────────── */

static int cmp_u32(const void *a, const void *b)
{
	uint32_t x = *(const uint32_t *)a, y = *(const uint32_t *)b;
	return (x > y) - (x < y);
}

void edge_signal_from_pcs(const unsigned long *pcs, int npc,
                          struct prog_signal *out)
{
	out->count = 0;
	out->capacity = npc > 0 ? npc : 16;
	out->edges = malloc(out->capacity * sizeof(uint32_t));

	/* Temporary bitmap to dedup within this program */
	uint8_t *seen = calloc(SIGNAL_MAP_SIZE / 8 + 1, 1);

	unsigned long prev = 0;
	for (int i = 0; i < npc; i++) {
		uint32_t edge = edge_hash(pcs[i], prev);
		prev = pcs[i];

		/* Dedup: only add each edge once per program */
		if (seen[edge / 8] & (1 << (edge % 8))) continue;
		seen[edge / 8] |= (1 << (edge % 8));

		if (out->count >= out->capacity) {
			out->capacity *= 2;
			out->edges = realloc(out->edges, out->capacity * sizeof(uint32_t));
		}
		out->edges[out->count++] = edge;
	}
	free(seen);

	qsort(out->edges, out->count, sizeof(uint32_t), cmp_u32);
}

void prog_signal_free(struct prog_signal *s)
{
	free(s->edges);
	s->edges = NULL;
	s->count = 0;
}

/* ─── Corpus ──────────────────────────────────────────────────────────────── */

void signal_corpus_init(struct signal_corpus *c)
{
	memset(c, 0, sizeof(*c));
	c->capacity = 256;
	c->entries = calloc(c->capacity, sizeof(struct signal_corpus_entry));
}

void signal_corpus_free(struct signal_corpus *c)
{
	for (int i = 0; i < c->count; i++)
		prog_signal_free(&c->entries[i].sig);
	free(c->entries);
	memset(c, 0, sizeof(*c));
}

int edge_signal_new_count(struct signal_corpus *corpus, struct prog_signal *sig)
{
	int new = 0;
	for (int i = 0; i < sig->count; i++) {
		uint32_t e = sig->edges[i];
		if (corpus->max_signal.map[e] == 0)
			new++;
	}
	return new;
}

int signal_corpus_add(struct signal_corpus *c, int prog_id,
                      struct prog_signal *sig)
{
	/* Count new edges */
	int new_edges = edge_signal_new_count(c, sig);
	if (new_edges == 0) {
		prog_signal_free(sig);
		return 0; /* no new signal, reject */
	}

	/* Update global max signal */
	for (int i = 0; i < sig->count; i++) {
		uint32_t e = sig->edges[i];
		if (c->max_signal.map[e] < 255)
			c->max_signal.map[e]++;
		if (c->max_signal.map[e] == 1)
			c->max_signal.total_edges++;
	}

	/* Add to corpus */
	if (c->count >= c->capacity) {
		c->capacity *= 2;
		c->entries = realloc(c->entries, c->capacity * sizeof(struct signal_corpus_entry));
	}
	struct signal_corpus_entry *e = &c->entries[c->count++];
	e->prog_id = prog_id;
	e->sig = *sig; /* transfer ownership */
	e->unique_edges = new_edges;

	return 1;
}

int signal_corpus_minimize(struct signal_corpus *c)
{
	if (c->count <= 1) return 0;

	/* Rebuild: which edges are covered by how many programs */
	uint16_t *edge_refcount = calloc(SIGNAL_MAP_SIZE, sizeof(uint16_t));
	for (int i = 0; i < c->count; i++)
		for (int j = 0; j < c->entries[i].sig.count; j++) {
			uint32_t e = c->entries[i].sig.edges[j];
			if (edge_refcount[e] < 65535) edge_refcount[e]++;
		}

	/* Mark programs for removal: those with 0 unique edges */
	int removed = 0;
	for (int i = 0; i < c->count; i++) {
		int has_unique = 0;
		for (int j = 0; j < c->entries[i].sig.count; j++) {
			if (edge_refcount[c->entries[i].sig.edges[j]] == 1) {
				has_unique = 1;
				break;
			}
		}
		if (!has_unique) {
			/* Remove: decrement refcounts and compact */
			for (int j = 0; j < c->entries[i].sig.count; j++)
				edge_refcount[c->entries[i].sig.edges[j]]--;
			prog_signal_free(&c->entries[i].sig);
			c->entries[i] = c->entries[--c->count];
			i--; /* re-check swapped entry */
			removed++;
		}
	}

	/* Update unique_edges counts */
	for (int i = 0; i < c->count; i++) {
		int unique = 0;
		for (int j = 0; j < c->entries[i].sig.count; j++)
			if (edge_refcount[c->entries[i].sig.edges[j]] == 1) unique++;
		c->entries[i].unique_edges = unique;
	}

	free(edge_refcount);
	return removed;
}

int signal_corpus_total_edges(struct signal_corpus *c)
{
	return c->max_signal.total_edges;
}

int signal_corpus_size(struct signal_corpus *c)
{
	return c->count;
}

File: vock/src/fuzz/signal_edge.rs
//! Edge-based signal tracking + corpus minimization (port of fuzz/signal_edge.c).
//!
//! Signal = hash(PC ^ prev_PC) for each consecutive PC pair in a KCOV trace.
//! The corpus keeps only programs that contribute at least one unique edge;
//! minimization removes programs whose edges are all covered by others.
//!
//! This mirrors the standalone `signal_edge_test` engine. It is not on the
//! main fuzz loop's hot path (fuzz.c uses the simpler `signal` set), but is
//! ported here in full for parity with the C sources.
#![allow(dead_code)]

pub const SIGNAL_MAP_BITS: u32 = 16;
pub const SIGNAL_MAP_SIZE: usize = 1 << SIGNAL_MAP_BITS;

/// Global max signal, union of all edges ever seen.
pub struct EdgeSignal {
    pub map: Vec<u8>, // hit count per edge bucket (SIGNAL_MAP_SIZE)
    pub total_edges: u32,
}

impl EdgeSignal {
    fn new() -> EdgeSignal {
        EdgeSignal {
            map: vec![0u8; SIGNAL_MAP_SIZE],
            total_edges: 0,
        }
    }
}

/// Per-program signal, the sorted set of edges the program contributes.
pub struct ProgSignal {
    pub edges: Vec<u32>,
}

/// Fast mixing hash of `pc ^ prev_pc` reduced to the edge-map index space.
fn edge_hash(pc: u64, prev_pc: u64) -> u32 {
    let mut h = pc ^ prev_pc;
    h ^= h >> 33;
    h = h.wrapping_mul(0xff51afd7ed558ccd);
    h ^= h >> 33;
    h = h.wrapping_mul(0xc4ceb9fe1a85ec53);
    h ^= h >> 33;
    (h & (SIGNAL_MAP_SIZE as u64 - 1)) as u32
}

/// Compute the edge signal from a raw KCOV PC array (deduped within-program,
/// sorted ascending).
pub fn edge_signal_from_pcs(pcs: &[u64]) -> ProgSignal {
    let mut seen = vec![0u8; SIGNAL_MAP_SIZE / 8 + 1];
    let mut edges: Vec<u32> = Vec::new();
    let mut prev: u64 = 0;
    for &pc in pcs {
        let edge = edge_hash(pc, prev);
        prev = pc;
        let (byte, bit) = ((edge / 8) as usize, (edge % 8) as u8);
        if seen[byte] & (1 << bit) != 0 {
            continue;
        }
        seen[byte] |= 1 << bit;
        edges.push(edge);
    }
    edges.sort_unstable();
    ProgSignal { edges }
}

pub struct SignalCorpusEntry {
    pub prog_id: i32,
    pub sig: ProgSignal,
    pub unique_edges: i32,
}

pub struct SignalCorpus {
    pub entries: Vec<SignalCorpusEntry>,
    pub max_signal: EdgeSignal,
}

impl SignalCorpus {
    pub fn new() -> SignalCorpus {
        SignalCorpus {
            entries: Vec::new(),
            max_signal: EdgeSignal::new(),
        }
    }

    /// How many edges in `sig` are new vs the global max signal.
    pub fn new_count(&self, sig: &ProgSignal) -> i32 {
        let mut new = 0;
        for &e in &sig.edges {
            if self.max_signal.map[e as usize] == 0 {
                new += 1;
            }
        }
        new
    }

    /// Add a program to the corpus if it contributes new signal.
    /// Returns true if added.
    pub fn add(&mut self, prog_id: i32, sig: ProgSignal) -> bool {
        let new_edges = self.new_count(&sig);
        if new_edges == 0 {
            return false; // no new signal, reject
        }
        for &e in &sig.edges {
            let e = e as usize;
            if self.max_signal.map[e] < 255 {
                self.max_signal.map[e] += 1;
            }
            if self.max_signal.map[e] == 1 {
                self.max_signal.total_edges += 1;
            }
        }
        self.entries.push(SignalCorpusEntry {
            prog_id,
            sig,
            unique_edges: new_edges,
        });
        true
    }

    /// Remove programs whose signal is a subset of others. Returns the count
    /// of removed programs.
    pub fn minimize(&mut self) -> i32 {
        if self.entries.len() <= 1 {
            return 0;
        }
        let mut edge_refcount = vec![0u16; SIGNAL_MAP_SIZE];
        for e in &self.entries {
            for &edge in &e.sig.edges {
                if edge_refcount[edge as usize] < u16::MAX {
                    edge_refcount[edge as usize] += 1;
                }
            }
        }

        let mut removed = 0;
        let mut i = 0;
        while i < self.entries.len() {
            let has_unique = self.entries[i]
                .sig
                .edges
                .iter()
                .any(|&edge| edge_refcount[edge as usize] == 1);
            if !has_unique {
                for &edge in &self.entries[i].sig.edges {
                    edge_refcount[edge as usize] -= 1;
                }
                // swap-remove and re-check the swapped-in entry (i unchanged)
                self.entries.swap_remove(i);
                removed += 1;
            } else {
                i += 1;
            }
        }

        for e in self.entries.iter_mut() {
            let unique = e
                .sig
                .edges
                .iter()
                .filter(|&&edge| edge_refcount[edge as usize] == 1)
                .count() as i32;
            e.unique_edges = unique;
        }
        removed
    }

    pub fn total_edges(&self) -> i32 {
        self.max_signal.total_edges as i32
    }

    pub fn size(&self) -> i32 {
        self.entries.len() as i32
    }
}
